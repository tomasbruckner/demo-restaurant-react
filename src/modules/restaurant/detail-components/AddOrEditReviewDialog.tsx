import React, { PureComponent } from 'react';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import Button from 'modules/common/components/controls/Button';
import AddCircleOutline from '@material-ui/icons/AddCircleOutline';
import FormDialog from 'modules/common/components/dialogs/FormDialog';
import { KeyboardDatePicker } from '@material-ui/pickers';
import Rating from 'modules/common/components/controls/Rating';
import Input from 'modules/common/components/controls/Input';
import styled from 'styled-components/macro';
import { withStyles } from '@material-ui/core';
import { reviewCreateAsync, reviewEditAsync } from 'core/redux/ducks/ReviewDuck';
import { connect, ConnectedProps } from 'react-redux';
import { ReviewResponse } from 'services/reviewService';

type Props = {
  restaurantId: number;
  defaultRestaurantValues?: ReviewResponse;
  hideButton?: boolean;
  onOpenToggle?: () => void;
  isControlledOpen?: boolean;
};

const StyledDiv = styled.div`
  margin-top: 16px;
`;

const StyledButton = withStyles({
  root: {
    margin: '16px',
  },
})(Button);

const connector = connect(null, {
  dispatchCreateReview: reviewCreateAsync,
  dispatchEditReview: reviewEditAsync,
});

const StyledInput = withStyles(theme => ({
  root: {
    margin: theme.spacing(1),
  },
}))(Input);

export class AddOrEditReviewDialog extends PureComponent<
  Props & ConnectedProps<typeof connector> & WithTranslation
> {
  state = {
    isOpen: false,
    selectedDate: new Date(),
  };

  onOpenToggle = () => {
    const { isOpen } = this.state;
    const { onOpenToggle } = this.props;

    if (onOpenToggle) {
      onOpenToggle();
    }

    this.setState({ isOpen: !isOpen });
  };

  getIsoDate = (date: string) => {
    const d = new Date(`${date} 13:00`);

    return d.toISOString();
  };

  onSubmit = (event: any) => {
    const {
      dispatchCreateReview,
      dispatchEditReview,
      restaurantId,
      defaultRestaurantValues,
    } = this.props;
    const { date, rating, title, review } = event.target.elements;

    if (defaultRestaurantValues) {
      const { reviewId } = defaultRestaurantValues;
      dispatchEditReview({
        data: {
          dateOfVisit: this.getIsoDate(date.value),
          reviewRatingId: rating.value,
          title: title.value,
          text: review.value,
          restaurantId,
        },
        reviewId,
      });
    } else {
      dispatchCreateReview({
        dateOfVisit: this.getIsoDate(date.value),
        reviewRatingId: rating.value,
        title: title.value,
        text: review.value,
        restaurantId,
      });
    }

    this.onOpenToggle();
  };

  onDateChange = (date: any) => {
    this.setState({ selectedDate: date });
  };

  render() {
    const { isControlledOpen, hideButton, defaultRestaurantValues, t } = this.props;
    const { isOpen, selectedDate } = this.state;

    return (
      <>
        {hideButton ? null : (
          <StyledButton
            variant="outlined"
            color="primary"
            startIcon={<AddCircleOutline />}
            data-cy="review-dialog-open"
            onClick={this.onOpenToggle}
          >
            {t('component.add-review-dialog.buttons.add')}
          </StyledButton>
        )}
        <FormDialog
          title={t('component.add-review-dialog.title')}
          submitButtonText={t('component.add-review-dialog.submit')}
          cancelButtonText={t('component.add-review-dialog.cancel')}
          onClose={this.onOpenToggle}
          isOpen={isControlledOpen !== undefined ? isControlledOpen : isOpen}
          onSubmit={this.onSubmit}
        >
          <KeyboardDatePicker
            disableToolbar
            variant="inline"
            format="MM/dd/yyyy"
            margin="none"
            id="date-picker-inline"
            name="date"
            label={t('component.add-review-dialog.date-of-visit')}
            value={selectedDate}
            defaultValue={defaultRestaurantValues ? defaultRestaurantValues.dateOfVisit : undefined}
            onChange={this.onDateChange}
          />
          <StyledDiv>
            <Rating
              name="rating"
              size="large"
              data-cy="review-dialog-rating"
              defaultValue={defaultRestaurantValues ? defaultRestaurantValues.reviewRatingId : 5}
            />
          </StyledDiv>
          <StyledInput
            id="idAddReviewDialogTitle"
            label={t('component.add-review-dialog.review-title')}
            name="title"
            required
            data-cy="review-dialog-title"
            defaultValue={defaultRestaurantValues ? defaultRestaurantValues.title : undefined}
            fullWidth
          />
          <StyledInput
            id="idAddReviewDialogReview"
            label={t('component.add-review-dialog.review')}
            name="review"
            data-cy="review-dialog-review"
            required
            defaultValue={defaultRestaurantValues ? defaultRestaurantValues.text : undefined}
            fullWidth
            rows={8}
            multiline
          />
        </FormDialog>
      </>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(AddOrEditReviewDialog));
