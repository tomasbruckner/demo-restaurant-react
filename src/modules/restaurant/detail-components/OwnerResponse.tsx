import React, { PureComponent } from 'react';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import Typography from 'modules/common/components/controls/Typography';
import Divider from 'modules/common/components/controls/Divider';
import { withStyles } from '@material-ui/core';
import styled from 'styled-components/macro';
import SimpleDialog from 'modules/common/components/dialogs/SimpleDialog';
import EditIcon from '@material-ui/icons/Edit';
import DeleteIcon from '@material-ui/icons/Delete';
import Button from 'modules/common/components/controls/Button';
import AddOrEditReviewReplyDialog from '../owned-restaurants-components/AddOrEditReviewReplyDialog';
import { ReviewResponse } from 'services/reviewService';
import { reviewReplyDeleteAsync } from '../../../core/redux/ducks/ReviewDuck';
import { connect, ConnectedProps } from 'react-redux';

const StyledDivider = withStyles({
  root: {
    marginLeft: '16px',
    marginRight: '16px',
  },
})(Divider);

const StyledWrapper = styled.div`
  display: flex;
  margin: 16px;
`;

const connector = connect(null, {
  dispatchDeleteReviewReply: reviewReplyDeleteAsync,
});

type Props = {
  review: ReviewResponse;
  canEdit?: boolean;
};

export class OwnerResponse extends PureComponent<
  Props & ConnectedProps<typeof connector> & WithTranslation
> {
  state = {
    isDeleteOpen: false,
    isEditOpen: false,
  };

  onToggleDelete = () => {
    const { isDeleteOpen } = this.state;

    this.setState({ isDeleteOpen: !isDeleteOpen });
  };

  onToggleEdit = () => {
    const { isEditOpen } = this.state;

    this.setState({ isEditOpen: !isEditOpen });
  };

  onDelete = async () => {
    const {
      dispatchDeleteReviewReply,
      review: { ownerResponse },
    } = this.props;

    if (ownerResponse && ownerResponse.reviewReplyId) {
      await dispatchDeleteReviewReply(ownerResponse.reviewReplyId);
    }
    this.onToggleDelete();
  };

  render() {
    const {
      review: { ownerResponse, restaurantId },
      canEdit,
      t,
    } = this.props;
    const { isEditOpen, isDeleteOpen } = this.state;

    if (!ownerResponse) {
      return null;
    }

    const { text, reviewId, reviewReplyId } = ownerResponse;

    return (
      <StyledWrapper>
        <StyledDivider orientation="vertical" flexItem />
        <div>
          <Typography variant="body1">{t('component.owner-response.title')}</Typography>
          <Typography variant="subtitle2" color="textSecondary">
            {text}
          </Typography>
          {canEdit ? (
            <span>
              <Button asIcon onClick={this.onToggleEdit}>
                <EditIcon />
              </Button>
              <Button asIcon onClick={this.onToggleDelete}>
                <DeleteIcon />
              </Button>
              <SimpleDialog
                onClose={this.onToggleDelete}
                onSubmit={this.onDelete}
                isOpen={isDeleteOpen}
                text={t('component.owner-response.delete-text')}
                title={t('component.owner-response.delete-title')}
              />
              <AddOrEditReviewReplyDialog
                restaurantId={restaurantId}
                isControlledOpen={isEditOpen}
                onToggle={this.onToggleEdit}
                hideButton
                defaultOwnerResponse={text}
                reviewReplyId={reviewReplyId}
                reviewId={reviewId}
              />
            </span>
          ) : null}
        </div>
      </StyledWrapper>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(OwnerResponse));
