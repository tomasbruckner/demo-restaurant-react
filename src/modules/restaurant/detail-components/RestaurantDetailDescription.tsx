import React, { PureComponent } from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';
import Typography from 'modules/common/components/controls/Typography';
import { Card, CardContent, withStyles } from '@material-ui/core';
import { I18N_CORE } from 'core/i18n/i18n';
import FileUploadDialog from 'modules/common/components/file-upload/FileUploadDialog';
import { API_ENDPOINT } from 'config/config';
import Avatar from '@material-ui/core/Avatar';
import RestaurantMenuIcon from '@material-ui/icons/RestaurantMenu';
import ImageService from 'services/imageService';
import { connect, ConnectedProps } from 'react-redux';
import { restaurantDetailAsync } from 'core/redux/ducks/RestaurantDuck';
import { RestaurantResponse } from '../../../services/restaurantService';
import { RootState } from '../../../core/redux/store';
import { getUserFromState } from '../../../core/redux/ducks/UserDuck';
import { Role } from '../../../utils/constants';

const StyledAvatar = withStyles({
  root: {
    height: '100%',
    width: '100%',
  },
})(Avatar);

const StyledMenuIcon = withStyles({
  root: {
    height: '100%',
    width: '100%',
  },
})(RestaurantMenuIcon);

type Props = {
  restaurant: RestaurantResponse;
};

const connector = connect(
  (state: RootState) => {
    return {
      user: getUserFromState(state),
    };
  },
  {
    dispatchGetRestaurant: restaurantDetailAsync,
  },
);

export class RestaurantDetailDescription extends PureComponent<
  Props & WithTranslation & ConnectedProps<typeof connector>
> {
  state = {
    isDialogOpen: false,
  };

  onToggleUploadDialog = () => {
    const { isDialogOpen } = this.state;

    this.setState({ isDialogOpen: !isDialogOpen });
  };

  onUploadFile = async (file: File) => {
    const {
      dispatchGetRestaurant,
      restaurant: { restaurantId },
    } = this.props;
    const formData = new FormData();
    formData.append('file', file);

    this.onToggleUploadDialog();
    await ImageService.uploadRestaurantImage({ data: formData, restaurantId });
    dispatchGetRestaurant(restaurantId);
  };

  render() {
    const {
      restaurant: { imageLink, description, ownerId },
      user,
      t,
    } = this.props;
    const { isDialogOpen } = this.state;
    const canEdit = user && (user.roleId === Role.Admin || user.userId === ownerId);

    return (
      <Card
        variant="outlined"
        style={{ display: 'flex', flexDirection: 'row', alignItems: 'center' }}
      >
        <StyledAvatar
          src={`${API_ENDPOINT}${imageLink}`}
          variant="square"
          onClick={canEdit ? this.onToggleUploadDialog : undefined}
          style={{
            height: '100%',
            maxHeight: '250px',
            maxWidth: '250px',
            cursor: canEdit ? 'pointer' : '',
          }}
        >
          <StyledMenuIcon />
        </StyledAvatar>
        {canEdit && (
          <FileUploadDialog
            onSubmit={this.onUploadFile}
            isOpen={isDialogOpen}
            onClose={this.onToggleUploadDialog}
          />
        )}
        <CardContent>
          <Typography variant="h5" component="h1">
            {t('component.restaurant-detail-description.title')}
          </Typography>
          <Typography variant="body2" component="p">
            {description}
          </Typography>
        </CardContent>
      </Card>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(RestaurantDetailDescription));
