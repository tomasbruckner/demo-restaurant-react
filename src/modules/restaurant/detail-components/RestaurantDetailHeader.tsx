import React, { PureComponent } from 'react';
import LocationOnIcon from '@material-ui/icons/LocationOn';
import CallIcon from '@material-ui/icons/Call';
import ComputerIcon from '@material-ui/icons/Computer';
import NorthEastIcon from '@material-ui/icons/CallMade';
import RatingWrapper from '../common-components/RatingWrapper';
import styled from 'styled-components/macro';
import Typography from 'modules/common/components/controls/Typography';
import Divider from 'modules/common/components/controls/Divider';
import { withStyles } from '@material-ui/core';
import RestaurantService, { RestaurantResponse } from 'services/restaurantService';
import SimpleDialog from 'modules/common/components/dialogs/SimpleDialog';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import { RouteComponentProps, withRouter } from 'react-router';
import { RESTAURANTS } from 'core/router/urls';
import RestaurantAddOrEditDialog from '../owned-restaurants-components/AddOrEditRestaurantDialog';
import { RootState } from '../../../core/redux/store';
import { getUserFromState } from '../../../core/redux/ducks/UserDuck';
import { connect, ConnectedProps } from 'react-redux';
import { Role } from '../../../utils/constants';

const StyledLink = styled.a`
  color: black;
  margin-left: 16px;
  margin-right: 16px;
  text-decoration: none;
  &:visited {
    color: black;
  }

  &:active {
    color: black;
  }

  &:hover {
    color: ${p => p.theme.primary};
  }

  &:link {
    color: black;
  }
`;

const StyledDivider = withStyles({
  root: {
    marginLeft: '16px',
    marginRight: '16px',
  },
})(Divider);

const StyledTypography = withStyles({
  root: {
    marginLeft: '8px',
    marginRight: '8px',
  },
})(Typography);

const StyledSpan = styled.span`
  display: flex;
  align-items: center;
`;

const StyledDiv = styled.div`
  display: flex;
  align-items: center;
`;

const StyledMarginDiv = styled.div`
  margin-bottom: 16px;
`;

const connector = connect((state: RootState) => {
  return {
    user: getUserFromState(state),
  };
});

type Props = {
  restaurant: RestaurantResponse;
};

export class RestaurantDetailHeader extends PureComponent<
  Props & ConnectedProps<typeof connector> & WithTranslation & RouteComponentProps
> {
  state = {
    isEditOpen: false,
    isDeleteOpen: false,
  };

  onDelete = async () => {
    const {
      restaurant: { restaurantId },
      history,
    } = this.props;

    await RestaurantService.delete(restaurantId);
    history.push(RESTAURANTS);
  };

  onToggleEdit = () => {
    const { isEditOpen } = this.state;

    this.setState({ isEditOpen: !isEditOpen });
  };

  onToggleDelete = () => {
    const { isDeleteOpen } = this.state;

    this.setState({ isDeleteOpen: !isDeleteOpen });
  };

  render() {
    const { restaurant, user, t } = this.props;
    const { averageRating, address, contactPhone, numberOfReviews = 0, websiteUrl, name, ownerId } =
      restaurant || {};
    const { isEditOpen, isDeleteOpen } = this.state;

    return (
      <StyledMarginDiv>
        <StyledDiv>
          <StyledTypography variant="h4" display="inline">
            {`${averageRating}`.substring(0, 3)}
          </StyledTypography>
          <RatingWrapper
            averageRating={averageRating}
            numberOfReviews={numberOfReviews}
            canEdit={user!.roleId === Role.Admin || user!.userId === ownerId}
            onDelete={this.onToggleDelete}
            onEdit={this.onToggleEdit}
          />
          <SimpleDialog
            onClose={this.onToggleDelete}
            onSubmit={this.onDelete}
            isOpen={isDeleteOpen}
            text={t('component.restaurant-detail-header.delete-text', { name })}
            title={t('component.restaurant-detail-header.delete-title')}
          />
          <RestaurantAddOrEditDialog
            defaultRestaurantValues={restaurant}
            isControlledOpen={isEditOpen}
            onOpenToggle={this.onToggleEdit}
            hideButton
          />
        </StyledDiv>
        <StyledDiv>
          <StyledSpan>
            <LocationOnIcon />
            <StyledTypography display="inline">{address}</StyledTypography>
          </StyledSpan>
          <StyledDivider orientation="vertical" flexItem />
          <StyledSpan>
            <CallIcon />
            <StyledTypography display="inline">{contactPhone}</StyledTypography>
          </StyledSpan>
          <StyledDivider orientation="vertical" flexItem />
          <StyledSpan>
            <ComputerIcon />
            <Typography component={StyledLink} href={websiteUrl} gutterBottom>
              <u>{websiteUrl}</u>
            </Typography>
            <NorthEastIcon />
          </StyledSpan>
        </StyledDiv>
      </StyledMarginDiv>
    );
  }
}

export default withTranslation(I18N_CORE)(withRouter(connector(RestaurantDetailHeader)));
