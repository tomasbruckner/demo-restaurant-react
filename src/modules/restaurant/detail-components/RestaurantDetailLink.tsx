import Link from 'modules/common/utils/Link';
import React, { PureComponent } from 'react';

export class RestaurantDetailLink extends PureComponent<{
  restaurantName: string;
  restaurantId: number;
}> {
  render() {
    const { restaurantName, restaurantId } = this.props;

    return <Link to={`restaurant/${restaurantId}/${restaurantName}`}>{restaurantName}</Link>;
  }
}

export default RestaurantDetailLink;
