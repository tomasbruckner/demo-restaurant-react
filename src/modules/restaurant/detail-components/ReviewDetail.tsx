import React, { PureComponent } from 'react';
import { I18N_CORE } from 'core/i18n/i18n';
import { Trans, WithTranslation, withTranslation } from 'react-i18next';
import { Card, CardContent, withStyles } from '@material-ui/core';
import Typography from 'modules/common/components/controls/Typography';
import { formatDate } from 'utils/commonUtils';
import Rating from 'modules/common/components/controls/Rating';
import { ReviewResponse } from 'services/reviewService';
import ThumbUpAltIcon from '@material-ui/icons/ThumbUp';
import ThumbDownAltIcon from '@material-ui/icons/ThumbDown';
import styled from 'styled-components/macro';
import { getRestaurantDetail } from 'core/router/urls';
import Link from 'modules/common/utils/Link';

const StyledThumpUp = withStyles({
  root: {
    color: 'green',
    marginLeft: '16px',
  },
})(ThumbUpAltIcon);

const StyledThumpDown = withStyles({
  root: {
    color: 'red',
    marginLeft: '16px',
  },
})(ThumbDownAltIcon);

const StyledTypographyFlex = withStyles({
  root: {
    display: 'flex',
    alignItems: 'center',
  },
})(Typography);

const StyledStrong = styled.strong`
  margin-right: 8px;
`;

const StyledTypography = withStyles({
  root: {
    margin: '16px',
  },
})(Typography);

const StyledLink = styled(Link)`
  color: black;
  text-decoration: none;
  &:visited {
    color: black;
  }

  &:active {
    color: black;
  }

  &:hover {
    color: black;
  }

  &:link {
    color: black;
  }
`;

const StyledRating = withStyles({
  root: {
    margin: '16px',
    marginBottom: 0,
  },
})(Rating);

const StyledCard = withStyles({
  root: {
    margin: 0,
  },
})(Card);

type Props = { positive?: boolean; data?: ReviewResponse; withRestaurantLink?: boolean };

export class ReviewDetail extends PureComponent<Props & WithTranslation> {
  render() {
    const { data, t, positive, withRestaurantLink } = this.props;

    if (!data) {
      return (
        <Card variant="outlined">
          <CardContent>
            <StyledTypography>
              {t(
                positive
                  ? 'component.review-detail.missing-positive'
                  : 'component.review-detail.missing-negative',
              )}
            </StyledTypography>
          </CardContent>
        </Card>
      );
    }

    const {
      owner: { firstName, lastName },
      reviewRatingId,
      title,
      text,
      created,
      dateOfVisit,
      restaurantId,
      restaurantName,
    } = data;
    const user = `${firstName} ${lastName}`;

    return (
      <StyledCard variant="outlined">
        <CardContent>
          {withRestaurantLink && (
            <Typography
              variant="h5"
              bottomGutter
              data-cy="review-detail-restaurant-link"
              component={StyledLink}
              to={getRestaurantDetail(restaurantId)}
            >
              {restaurantName}
            </Typography>
          )}
          <StyledTypographyFlex variant="subtitle2">
            <Trans
              i18nKey={
                positive
                  ? 'component.review-detail.reviewer-positive'
                  : positive === false
                  ? 'component.review-detail.reviewer-negative'
                  : 'component.review-detail.reviewer'
              }
              values={{ user }}
              components={{ strong: <StyledStrong /> }}
            />
            {positive ? <StyledThumpUp /> : positive === false ? <StyledThumpDown /> : null}
          </StyledTypographyFlex>
          <Typography variant="subtitle2" color="textSecondary">
            {formatDate(created)}
          </Typography>
          <StyledRating precision={0.5} value={reviewRatingId} readOnly />
          <StyledTypography variant="h5">{title}</StyledTypography>
          <StyledTypography variant="body2" color="textSecondary">
            <i>{`"${text}"`}</i>
          </StyledTypography>
          <StyledTypography variant="body2">
            <i>
              <Trans
                i18nKey="component.review-detail.date-of-visit"
                values={{ date: formatDate(dateOfVisit) }}
                components={{ strong: <strong /> }}
              />
            </i>
          </StyledTypography>
        </CardContent>
      </StyledCard>
    );
  }
}

export default withTranslation(I18N_CORE)(ReviewDetail);
