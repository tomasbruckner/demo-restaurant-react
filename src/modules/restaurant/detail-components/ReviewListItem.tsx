import React, { PureComponent } from 'react';
import { Trans, withTranslation, WithTranslation } from 'react-i18next';
import { I18N_CORE } from 'core/i18n/i18n';
import { ReviewResponse } from 'services/reviewService';
import Typography from 'modules/common/components/controls/Typography';
import RatingWrapper from '../common-components/RatingWrapper';
import { Card, CardContent, withStyles } from '@material-ui/core';
import OwnerResponse from './OwnerResponse';
import { formatDate } from 'utils/commonUtils';
import UserProfileImage from './UserProfileImage';
import SimpleDialog from '../../common/components/dialogs/SimpleDialog';
import AddOrEditReviewDialog from './AddOrEditReviewDialog';
import { reviewDeleteAsync } from 'core/redux/ducks/ReviewDuck';
import { connect, ConnectedProps } from 'react-redux';
import { RootState } from 'core/redux/store';
import { getUserFromState } from 'core/redux/ducks/UserDuck';
import { Role } from '../../../utils/constants';

const StyledTypography = withStyles({
  root: {
    marginBottom: '32px',
    marginRight: '16px',
    marginTop: '16px',
  },
})(Typography);

export type Props = {
  review: ReviewResponse;
};

const connector = connect(
  (state: RootState) => {
    return {
      user: getUserFromState(state),
    };
  },
  {
    dispatchReviewDelete: reviewDeleteAsync,
  },
);

const StyledCard = withStyles({
  root: {
    margin: '16px',
  },
})(Card);

export class ReviewListItem extends PureComponent<
  ConnectedProps<typeof connector> & Props & WithTranslation
> {
  state = {
    isDeleteOpen: false,
    isEditOpen: false,
  };

  onToggleDelete = () => {
    const { isDeleteOpen } = this.state;

    this.setState({ isDeleteOpen: !isDeleteOpen });
  };

  onToggleEdit = () => {
    const { isEditOpen } = this.state;

    this.setState({ isEditOpen: !isEditOpen });
  };

  onDelete = async () => {
    const {
      review: { reviewId, restaurantId },
      dispatchReviewDelete,
    } = this.props;

    await dispatchReviewDelete({ reviewId, restaurantId });
    this.onToggleDelete();
  };

  render() {
    const { review, user, t } = this.props;
    const { isEditOpen, isDeleteOpen } = this.state;
    const {
      owner: { firstName, lastName, imageLink },
      ownerResponse,
      dateOfVisit,
      title,
      text,
      restaurantId,
      created,
      reviewRatingId,
    } = review;

    return (
      <StyledCard variant="outlined">
        <CardContent style={{ display: 'flex' }}>
          <div>
            <StyledTypography variant="subtitle1">
              <strong>{`${firstName} ${lastName}`}</strong>
            </StyledTypography>
            <UserProfileImage imageLink={imageLink} />
          </div>
          <div>
            <RatingWrapper
              averageRating={reviewRatingId}
              text={t('component.review-list-item.reviewed-at', { date: formatDate(created) })}
              canEdit={
                user!.roleId === Role.Admin || (review && user!.userId === review.owner.userId)
              }
              onDelete={this.onToggleDelete}
              onEdit={this.onToggleEdit}
            />

            <SimpleDialog
              onClose={this.onToggleDelete}
              onSubmit={this.onDelete}
              isOpen={isDeleteOpen}
              text={t('component.review-list-item.delete-text')}
              title={t('component.review-list-item.delete-title')}
            />
            <AddOrEditReviewDialog
              hideButton
              onOpenToggle={this.onToggleEdit}
              isControlledOpen={isEditOpen}
              restaurantId={restaurantId}
              defaultRestaurantValues={review}
            />
            <Typography variant="h5">{title}</Typography>
            <Typography variant="body1">{text}</Typography>
            <Typography variant="subtitle1">
              {' '}
              <Trans
                i18nKey="component.review-list-item.date-of-visit"
                values={{ date: formatDate(dateOfVisit) }}
                components={{ strong: <strong /> }}
              />
            </Typography>
            <OwnerResponse
              review={review}
              canEdit={
                user!.roleId === Role.Admin ||
                (ownerResponse && user!.userId === ownerResponse.creatorId)
              }
            />
          </div>
        </CardContent>
      </StyledCard>
    );
  }
}

export default withTranslation(I18N_CORE)(connector(ReviewListItem));
