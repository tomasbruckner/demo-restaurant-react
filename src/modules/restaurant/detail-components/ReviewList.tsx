import React, { PureComponent } from 'react';
import ReviewListItem from './ReviewListItem';
import { I18N_CORE } from 'core/i18n/i18n';
import { WithTranslation, withTranslation } from 'react-i18next';
import Typography from 'modules/common/components/controls/Typography';
import { withStyles } from '@material-ui/core';
import { PagedDto } from 'types/commonTypes';
import { ReviewResponse } from 'services/reviewService';

type Props = {
  reviews: PagedDto<ReviewResponse>;
};

const StyledTypography = withStyles({
  root: {
    margin: '16px',
  },
})(Typography);

export class ReviewList extends PureComponent<Props & WithTranslation> {
  render() {
    const { reviews, t } = this.props;

    if (reviews.size === 0) {
      return (
        <StyledTypography variant="h5">
          <i>{t('page.restaurant-detail.no-reviews')}</i>
        </StyledTypography>
      );
    }

    return reviews.data.map(o => <ReviewListItem key={o.reviewId} review={o} />);
  }
}

export default withTranslation(I18N_CORE)(ReviewList);
