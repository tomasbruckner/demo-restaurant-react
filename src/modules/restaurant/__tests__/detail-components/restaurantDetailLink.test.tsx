import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { RestaurantDetailLink } from 'modules/restaurant/detail-components/RestaurantDetailLink';
import { fakeTranslationProps } from 'utils/testUtils';
import { BrowserRouter as Router } from 'react-router-dom';

describe('RestaurantDetailLink', () => {
  it('basic example', () => {
    const tree = renderer
      .create(
        <Router>
          <RestaurantDetailLink
            restaurantId={1}
            restaurantName="restaurant"
            {...fakeTranslationProps}
          />
        </Router>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
