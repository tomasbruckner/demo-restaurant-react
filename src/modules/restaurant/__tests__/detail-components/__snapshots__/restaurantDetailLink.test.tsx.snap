// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantDetailLink basic example 1`] = `
<a
  href="/restaurant/1/restaurant"
  onClick={[Function]}
>
  restaurant
</a>
`;
